pkg/loki/loki.go

```diff
@@ -0,0 +1,89 @@
+package loki
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/grafana/loki/pkg/logcli/client"
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/pingcap/errors"
+)
+
+const (
+	equalMatcher    = "|="
+	regexMatcher    = "|~"
+)
+
+type lokiClient struct {
+	cli       *client.Client
+	startTime time.Time
+}
+
+// NewLokiClient creates a client to query loki.
+func NewLokiClient(address, username, password string) *lokiClient {
+	return &lokiClient{
+		cli: &client.Client{
+			Address:  address,
+			Username: username,
+			Password: password,
+		},
+		startTime: time.Now(),
+	}
+}
+
+// FetchContainerLogs gets container logs from loki API. ns and containerName are used
+// to match the specific logs in loki. match is a string which you want to query
+// from loki, you can set isRegex to true to make it to be a regex match. nonMatch
+//  is a set of strings you don't want to match. lookbackTime is the query duration.
+func (c *lokiClient) FetchContainerLogs(ns, containerName, match string, nonMatch []string, lookbackTime time.Duration, isRegex bool) ([]string, int, error) {
+	if ns == "" {
+		return nil, 0, errors.New("namespace must be set")
+	}
+
+	if containerName == "" {
+		return nil, 0, errors.New("container name must be set")
+	}
+
+	if match == "" {
+		return nil, 0, errors.New("match query must be set")
+	}
+
+	var op string
+	if isRegex {
+		op = regexMatcher
+	} else {
+		op = equalMatcher
+	}
+
+	// Format the query to loki.
+	query := fmt.Sprintf(`{container_name="%s", namespace="%s"} %s"%s"`,
+		containerName, ns, op, match)
+
+	var nonEqual string
+	for _, v := range nonMatch {
+		nonEqual += fmt.Sprintf(` !="%s"`, v)
+	}
+	query += nonEqual
+
+	queryTo := time.Now()
+	queryFrom := queryTo.Add(-lookbackTime)
+	if queryFrom.Before(c.startTime) {
+		queryFrom = c.startTime
+	}
+
+	res, err := c.cli.QueryRange(query, 1000, queryFrom, queryTo, logproto.BACKWARD, 15*time.Second, true)
```

Comment:
What will be if there are more than 1000 results in this duration?

Reply:
If there are more than 1000 line of logs, it will return 1000. Making this parameter more than 1000 is not recommended. If allow more too many lines of logs, then it will cause a huge stress to loki itself.

Reply:
Should we handle this situation and make multi requests for each batch?

Reply:
I don't think we need. But maybe @mahjonp has some other ideas. IMO we just use logs to grep some key words like `error` or `panic`. Btw, the `lookbackTime` should not be too long, I don't think it can exceed `1d`, then it will too slow.

Reply:
this LGTM

Reply:
So the 1000 here is the filtering log limit or the scan log lines limit? if it's the first one, I think it's big enough

Reply:
> I don't think we need. But maybe @mahjonp has some other ideas. IMO we just use logs to grep some key words like `error` or `panic`. Btw, the `lookbackTime` should not be too long, I don't think it can exceed `1d`, then it will too slow.

Yes, it's usually small. Our case is running round by round, and each round isn't too long in normal.

Reply:
The limit of the results. Not the logs loki scans.